### PR TITLE
Language Server Protocol using atom-languageclient

### PR DESCRIPTION
**Pull request by keplersj**
Ref: https://github.com/Microsoft/vscode-tslint/issues/250

/cc: @egamma @damieng

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -1,156 +1,87 @@
...
+const { AutoLanguageClient } = require('atom-languageclient');
 
-// eslint-disable-next-line import/extensions, import/no-extraneous-dependencies
-import { CompositeDisposable, Task } from 'atom';
-import path from 'path';
-import fs from 'fs';
-import * as workerHelper from './workerHelper';
+class TSLintLanguageClient extends AutoLanguageClient {
+  getGrammarScopes() {
+    return atom.config.get('linter-tslint.javascriptSupport')
```
Comment (Arcanemagus) [inline lib/main.js]: Settings should always be observed, where would the best place to do that be here? `startServerProcess`? Or maybe in the constructor?

Comment (damieng) [reply lib/main.js]: Ideally they would but it gets complex fast - you'd need to handle open files of the type supported/no longer supported and send the appropriate didopen or didclose events.  It's a lot of complexity for uncommon scenario.  I'm also not sure that you can force many of the providers to re-evaluate the service endpoints that passed back these grammar scopes.

You might just be better off removing it - it was a last-minute workaround for people who need flow and typescript installed side-by-side.

File: lib/main.js [hunk 2]
```diff
@@ -1,156 +1,87 @@
...
+      configFile: path.join(this.projectPath, 'tslint.json'),
+      ignoreDefinitionFiles: atom.config.get('linter-tslint.ignoreTypings'),
+      exclude: [''],
+      validateWithDefaultConfig: !atom.config.get('linter.tslint.useLocalTslint'),
+      nodePath: process.execPath,
+      run: 'onType',
+      alwaysShowRuleFailuresAsWarnings: false,
+      autoFixOnSave: atom.config.get('linter-tslint.fixOnSave'),
+      trace: atom.config.get('core.debugLSP') ? { server: 'verbose' } : {},
+    }));
```
Comment (damieng) [inline lib/main.js]: Nice use of onRequest to implement workspace configuration!  Given that project specific settings in atom aren't a thing I wonder if we can do something to facilitate something similar with less effort just for our providers in lsp.

File: lib/main.js (hunk 1 above)
Comment (Arcanemagus) [reply lib/main.js]: Oh, I was talking about in general, there are far more uses of `atom.config.get` in this code.

File: lib/main.js (hunk 2 above)
Comment (damieng) [reply lib/main.js]: should also observe these config values and send a 'workspace/didChangeConfiguration' when that happens.

Comment (keplersj) [reply lib/main.js]: Awesome idea. Done!

File: lib/main.js (hunk 1 above)
Comment (keplersj) [reply lib/main.js]: Taken care of!

**Comment by keplersj** [comment]
Okay. I'm stumped. The package is communicating with the server I can see so in the console. It's only receiving empty "publishDiagnostics" notifications though. Any ideas @damieng @egamma?

**Comment by damieng** [comment]
The usual sequence is an initialization request and response followed by a didOpen followed by diagnostics for that file...  What are you seeing?

**Comment by keplersj** [comment]
I see this chain of events in the console after opening a TypeScript file:

1. `rpc.sendRequest initialize sending`
2. `rpc.sendRequest initialize received`
3. `rpc.sendNotification textDocument/didOpen`
4. `Server started`
5. `rpc.onRequest workspace/configuration`

When I save the file after opening I get:

1. `rpc.sendNotification textDocument/didSave`
2. `rpc.sendNotification workspace/didChangeWatchedFiles` (the actual file)
3. `rpc.sendNotification workspace/didChangeWatchedFiles` (`.git/index.lock`)
4. `rpc.onNotification` (`"window/logMessage"`: `"TSLint library loaded [...]"`)
5. `rpc.onNotification textDocument/publishDiagnostics` (empty array for correct file)
6. `rpc.onNotification textDocument/publishDiagnostics` (empty array for correct file)

**Comment by damieng** [comment]
If you then close and open another file do you get anything?  The initial
sequence looks a little unusual but it should still handle new open/close
requests after all that...

**Comment by keplersj** [comment]
Open TypeScript file:

1: `Server starting`
2: `starting child Node`
3: `rpc.sendNotification workspace/didChangeConfiguration`
4: `rpc.sendNotification workspace/didChangeConfiguration`
5: `rpc.sendNotification workspace/didChangeConfiguration`
6: `rpc.sendNotification workspace/didChangeConfiguration`
7: `rpc.sendNotification workspace/didChangeConfiguration`
8: `rpc.sendNotification workspace/didChangeConfiguration`
9: `rpc.sendRequest initialize sending`
10: `rpc.sendRequest initialize received`
11: `rpc.sendNotification textDocument/didOpen`
12: `Server started`
13: `rpc.onRequest workspace/configuration`

TypeScript file closed:

1: `rpc.sendNotification textDocument/didClose`
2: `rpc.onNotification textDocument/publishDiagnostics` (empty array, correct file)

TypeScriptReact file opened:

1: `rpc.sendNotification textDocument/didOpen`
2: `rpc.onRequest workspace/configuration`
3: `rpc.sendNotification textDocument/didChange`
4: `rpc.sendNotification textDocument/didChange`
5: `rpc.sendNotification textDocument/didChange`
6: `rpc.sendNotification textDocument/didChange`

TypeScriptReact file closed:

1: `rpc.sendNotification textDocument/didClose`
2: `rpc.onNotification textDocument/publishDiagnostics` (empty array, correct file)

Both the TypeScript and TypeScriptReact have purposeful errors in them that are detected in the CLI and the existing Atom implementation.

File: lib/main.js [hunk 3]
```diff
@@ -1,156 +1,98 @@
...
-        const results = await workerHelper.requestJob('lint', textEditor);
+    const configChanged = () => {
+      connection.didChangeConfiguration({ settings: getConfig() });
+    };
 
-        if (textEditor.getText() !== text) {
-          // Text has been modified since the lint was triggered, tell linter not to update
-          return null;
-        }
+    atom.config.observe('linter-tslint.javascriptSupport', configChanged);
```
Comment (Arcanemagus) [inline lib/main.js]: `atom.config.observe` returns a subscription, you should keep track of these all and dispose of them when the package's `deactivate` is called. Look at the original implementation for how to do this.

File: lib/main.js [hunk 4]
```diff
@@ -1,156 +1,98 @@
...
-        if (this.ignoreTypings && textEditor.getPath().toLowerCase().endsWith('.d.ts')) {
-          return [];
-        }
 
-        if (!this.worker) {
-          await waitOnIdle();
-        }
+    const getConfig = () => ({
+      enable: true,
+      jsEnable: atom.config.get('linter-tslint.javascriptSupport'),
```
Comment (Arcanemagus) [inline lib/main.js]: You shouldn't have _any_ calls to `atom.config.get`, the `atom.config.observe` you have below registers a callback that is called when first initialized and then any time the setting is changed. In that callback you should be updating a local cache of the setting values. In this function here you should be simply returning the current state of that cache.

File: lib/main.js [hunk 5]
```diff
@@ -1,156 +1,98 @@
...
+      [require.resolve('@keplersj/tslint-server/out/server')],
+      { cwd: projectPath, stdio: [null, null, null, 'ipc'] } // eslint-disable-line comma-dangle
     );
-
-    const createWorkerCallback = window.requestIdleCallback(() => {
-      this.worker = new Task(require.resolve('./worker.js'));
-      idleCallbacks.delete(createWorkerCallback);
+  }
+
+  preInitialization(connection) {
```
Comment (damieng) [inline lib/main.js]: I'd recommend moving most of this code - if not all of it - to the postInitialization hook as messages shouldn't be going back and forth until the initialization is complete.

The postInitialization function receives a server object which has a connection object on it as a property.

**Comment by egamma** [comment]
@keplersj 

> It's only receiving empty "publishDiagnostics" notifications though. Any ideas @damieng @egamma?

You can observe the protocol interaction in vscode by defining the setting `tslint.server.trace` to `verbose`. It will also trace why the diagnostics are empty (added in the latest version 1.0.7 of the extension).

Also keep in mind that the server is now using several proposed protocol extensions as described here https://github.com/Microsoft/language-server-protocol/blob/master/contributing.md

**Comment by keplersj** [comment]
I just gave this another go tonight and am happy to say, IT WORKS! Gonna need some polish and additional eyes. Should probably make use of [`typescript-eslint-parser`](https://github.com/eslint/typescript-eslint-parser) too.

File: lib/main.js (hunk 5 above)
Comment (keplersj) [reply lib/main.js]: Did this in my most recent commit. Thanks for the feedback!

**Comment by keplersj** [comment]
@egamma It is worth noting (particularly for https://github.com/Microsoft/vscode-tslint/issues/250) that the only thing I needed to modify in `vscode-tslint` was expose the server as a binary. No source code changes were needed.

**Comment by egamma** [comment]
@keplersj glad to see that this worked. What I meant in my comment https://github.com/Microsoft/vscode-tslint/issues/250 is that some additional features, like auto fixes will not work in Atom since they rely on custom messages, that currently only VS Code handles.

**Comment by keplersj** [comment]
Due to the on-going deprecation of TSLint and community-wide transition away from TSLint to ESLint, I am closing this Pull Request. I'll keep my fork up (in an archived state) for any passers-by interested in using a Language Service Provider in a Linter package for Atom.

**Pull request closed without merging**